[PM-3000] Add Server URLs to Account Switcher

rr-bw opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [ ] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

<!--Describe what the purpose of this PR is. For example: what bug you're fixing or what new feature you're adding-->

## Code changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
<!--Also refer to any related changes or PRs in other repositories-->

- **file.ext:** Description of what was changed and why

## Screenshots

<!--Required for any UI changes. Delete if not applicable-->

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

jlf0dev commented on libs/common/src/platform/misc/utils.ts:
```diff
@@ -263,6 +263,10 @@ export class Utils {
     return null;
   }
 
+  static removeVaultfromHostname(uriString: string): string {
```
nit (non-blocking): could you add jsdocs here with an example? I'd like to start adding these to all our utils

Hinton replied:
I don't believe this belongs to the utils.

jlf0dev reviewed (approved):
Looks good! Just had a small suggestion but its non-blocking

Hinton commented on apps/desktop/src/app/layout/account-switcher.component.ts:
```diff
@@ -165,11 +173,20 @@ export class AccountSwitcherComponent implements OnInit, OnDestroy {
+        userId: userId,
+      });
+
+      baseAccounts[userId].settings.region = Utils.removeVaultfromHostname(
+        Utils.getHostname(serverConfig.environment.vault)
+      );
```
A component shouldn't modify the state of accounts.

Hinton reviewed (changes requested):
We seem to be directly modifying accounts state in a component. This shouldn't happen.

Hinton reviewed (commented):
Apologies for my previous comments they were quite abrupt.

I believe the account switcher has gone through a series of unfortunate refactors coupling it tighter and tighter to various states. This is a significant code smell and I believe this PR worsens the situation by loading the server config from state.

The `StateService` should be viewed as an internal services that other DomainServices can rely on to store and fetch their domain. Inter-domain fetching should be strictly forbidden. This means `DomainServiceA` is not allowed to call `StateService` to fetch `DomainB`.

Concrete blockers as I view it:
- We're introducing `getServerConfig`, this should at a minimum go through the config service first.
- The `Utils.removeVaultFromHostname` seems potentially dangerous. You could have a self-hosted vault on `vault.company.com` which would now show up as `company.com`. Is that intentional?

The broader question is, why are we loading the config service in the first place? The user has EnvironmentUrls which should contain this information.

---

There are some quick wins that can be had in this component in my opinion.

- Remove SwitcherAccount, the component should define it's own model for the template. This decouples it from internal state models. (Check ActiveAccount)
- Remove `authenticationStatus` from `AccountProfile` it's weird that a state model contains a field only populated by a component. This can either be solved by introducing a new type using `& { authenticationStatus: boolean }`

Once we have a proper `AccountService` we should be able to remove direct calls to state service for these things.

rr-bw commented:
@Hinton Thanks for the very helpful feedback.

I've refactored quite a bit and I think this addresses all of your concerns and ensures this PR is not introducing any new calls to `stateService` from the component. Here is a summary of recent changes:

- I removed dependency on `getServerConfig(...)` entirely, and replaced it with two new methods in the `EnvironmentService`:
    - `getRegion(...)` - allows us to optionally pass in a `userId` to get the region via `StateService`
    - `getHost(...)` - uses the `getRegion(...)` method above to get the host name for US, EU, or Self-Hosted

- I refactored the `SwitcherAccount` class to an `InactiveAccount` intersection type (extends `ActiveAccount`). This way the component is only dealing with the data it actually needs (that is, we are no longer extending `Account`, and no longer adding `authenticationStatus` to `AccountProfile` via the component).
    - As a side note, I did some renaming here, from `SwitcherAccount` to `InactiveAccount`, because I believe this more clearly distinguishes the two types of accounts we are working with: `ActiveAccount` and `InactiveAccount`.

Let me know what you think. Thanks again!

Tagging @jlf0dev and @trmartin4 for visibility

Hinton commented on libs/common/src/platform/services/environment.service.ts:
```diff
@@ -283,6 +285,26 @@ export class EnvironmentService implements EnvironmentServiceAbstraction {
+      default:
+        // self-hosted host
+        return Utils.getHost(
+          (await this.stateService.getEnvironmentUrls(userId ? { userId: userId } : null)).webVault
```
I can't remember how we store urls, but do we need to handle `webVault` being null if the user only sets baseUrl?

Hinton reviewed (approved):
Nice, this looks significantly more maintainable. Thanks for iterating on this!

Hinton commented on libs/common/src/platform/services/environment.service.ts:
```diff
@@ -283,6 +285,27 @@ export class EnvironmentService implements EnvironmentServiceAbstraction {
+  async getHost(userId?: string) {
+    const region = await this.getRegion(userId ? userId : null);
+
+    let envUrls;
+
+    switch (region) {
+      case Region.US:
+        return RegionDomain.US;
+      case Region.EU:
+        return RegionDomain.EU;
+      default:
+        // self-hosted host
+        envUrls = await this.stateService.getEnvironmentUrls(userId ? { userId: userId } : null);
+        return Utils.getHost(envUrls.webVault || envUrls.base);
+    }
```
```suggestion
    switch (region) {
      case Region.US:
        return RegionDomain.US;
      case Region.EU:
        return RegionDomain.EU;
      default:
        // self-hosted host
        const envUrls = await this.stateService.getEnvironmentUrls(userId ? { userId: userId } : null);
        return Utils.getHost(envUrls.webVault || envUrls.base);
    }
```

rr-bw replied:
I did try that but I get an ESLint error. Would you like me to disable for this line?

![Screenshot 2023-10-16 at 9 56 35 AM](https://github.com/bitwarden/clients/assets/102181210/1d49be9f-539c-4f68-ac3f-99fe990fb1f4)

Hinton replied:
https://eslint.org/docs/latest/rules/no-case-declarations

The linting rule tells you that declarations in switch statements "leaks" into the parent scope. To avoid that it suggests wrapping clauses in blocks.

```ts
default: {
  // self-hosted host
  const envUrls = await this.stateService.getEnvironmentUrls(userId ? { userId: userId } : null);
  return Utils.getHost(envUrls.webVault || envUrls.base);
}
```

Hinton reviewed (approved):
Great improvement.

Hinton commented on apps/desktop/src/app/layout/account-switcher.component.html:
```diff
@@ -55,38 +56,40 @@
-            {{
-              (a.value.profile.authenticationStatus === authStatus.Unlocked ? "unlocked" : "locked")
+          <span class="email" aria-hidden="true">{{ account.value.email }}</span>
+          <span class="server" aria-hidden="true"
+            ><span class="sr-only"> / </span>{{ account.value.server }}</span
+          >
```
The formatting here seems a bit confusing.
```suggestion
          <span class="server" aria-hidden="true">
            <span class="sr-only"> / </span>{{ account.value.server }}
          </span>
```
